pyiceberg/table/__init__.py

```diff
@@ -1904,3 +1913,200 @@ def _generate_snapshot_id() -> int:
+        self._adds.append(new_field)
+        return self
+    
+    def remove_field(self, name: str) -> UpdateSpec:
+        added = self._name_to_added_field.get(name)
+        if added:
+            raise ValueError(f"Cannot delete newly added field {name}")
+        renamed = self._renames.get(name)
+        if renamed:
+            raise ValueError(f"Cannot rename and delete field: {name}")
+        field = self._name_to_field.get(name)
+        if not field:
+            raise ValueError(f"No such partition field: {name}")
+
+        self._deletes.add(field.field_id)
+        return self
+
+    def rename_field(self, name: str, new_name: str) -> UpdateSpec:
+        existing_field = self._name_to_field.get(new_name)
+        if existing_field and isinstance(existing_field.transform, VoidTransform):
+            return self.rename_field(name, name + "_" + existing_field.field_id)
+        added = self._name_to_added_field.get(name)
+        if added:
+            raise ValueError(f"Cannot rename recently added partitions")
+        field = self._name_to_field.get(name)
+        if not field:
+            raise ValueError(f"Cannot find partition field {name}")
+        if field.field_id in self._deletes:
+            raise ValueError(f"Cannot delete and rename partition field {name}")
+        self._renames[name] = new_name
+        return self
+    
+    def commit(self):
+        new_spec = self._apply()
+        updates = []
+        requirements = []
+        if self._table.metadata.default_spec_id != new_spec.spec_id:
+            if new_spec.spec_id not in self._table.specs():
+                spec_update = AddPartitionSpecUpdate(spec=new_spec)
+                updates.append(spec_update)
+                requirements.append(AssertLastAssignedPartitionId(last_assigned_partition_id=self._table.metadata.last_partition_id)) 
+
+            updates.append(SetDefaultSpecUpdate(spec_id=new_spec.spec_id))
+            requirements.append(AssertDefaultSpecId(default_spec_id=self._table.metadata.default_spec_id))        
+            self._table._do_commit(updates=updates, requirements=requirements)  # pylint: disable=W0212
+            
+
+    def _apply(self) -> PartitionSpec:
+        last_assigned_field_id = PARTITION_FIELD_ID_START - 1
+        partition_fields = []
+        partition_names = set([])
+
+        def _check_and_add_partition_name(schema: Schema, name: str, source_id: int, partition_names: Set[int]):
+            field = schema.find_field(name)
+            if source_id and field and field.field_id != source_id:
+                raise ValueError(f"Cannot create identity partition from a different field in the schema {name}")
+            elif field:
+                raise ValueError(f"Cannot create partition from name that exists in schema {name}")
+            if not name:
+                raise ValueError(f"Undefined empty/none name")
+            if name in partition_names:
```

Comment:
Unsure about this check. The `partition_names` is a `Set[int]` and the name is a `str`. Would be caught when writing tests